sound/soc/sof/ipc4-pcm.c

```diff
@@ -32,6 +35,236 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
+	dev_dbg(sdev->dev, "accuracy %d\n", spcm->accuracy);
+}
+
+/* the wclk is calculated by math interpolation and use it to match the real hw wclk */
+static inline void ipc4_match_hw_wclk(struct snd_sof_dev *sdev, u64 *wclk)
+{
+	u32 wclk_hw[] = {19200000, 20000000, 24576000, 24000000,
+		38400000, 48000000, 96000000};
+	int count = sizeof(wclk_hw);
+	int i;
+
+	for (i = 0; i < count; i++) {
+		if (*wclk >= wclk_hw[i] && *wclk < wclk_hw[i + 1])
+			break;
+	}
+
+	if (i >= count) {
+		dev_warn(sdev->dev, "warning can't find matched hw wclk %lld\n", *wclk);
+		return;
+	}
+
+	/* find nearest wclk */
+	if (*wclk > wclk_hw[i] + ((wclk_hw[i + 1] - wclk_hw[i]) >> 1))
+		*wclk = wclk_hw[i + 1];
+	else
+		*wclk = wclk_hw[i];
+}
+
+/*
+ * calculate stream start time and resolution of wclk in dsp
+ * The llp position in memory window is not the real position for a stream in two cases
+ * (1) dai starts before host in multiple pipeline cases. The llp counter works before
+ *  stream is started.
+ * (2) multiple streams with mixer. The llp counter is same for all streams.
+ *
+ * the solution is using stream_start_offset for each stream to save the llp counter
+ * when that stream is started. stream_start_offset is based on sample size, so need
+ * to convert it to sample count.
+ * the stream position = current llp value  - stream_start_offset
+ *
+ * to calculate stream start time in terms of wclk (wclk0), we interpolate the value
+ * stream_start_wclk			wclk1					wclk2
+ * stream_start_offsets*		llp1					llp2
+ *
+ * stream_start_wclk = wclk2 - (wclk1 -wclk2)/(llp1 -llp2) * (llp2  - stream_start_offsets)
+ */
+static void sof_ipc4_wclk_calibration_work(struct work_struct *work)
+{
+	struct sof_ipc4_pipeline_registers *ppl_reg = NULL;
+	struct sof_ipc4_llp_reading_slot *llp_slot = NULL;
+	struct sof_ipc4_fw_registers *fw_reg;
+	struct sof_ipc4_copier *host_copier = NULL;
+	struct sof_ipc4_copier *dai_copier = NULL;
+	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_pcm_stream *spcm;
+	struct snd_sof_dev *sdev;
+	struct snd_sof_dai *dai;
+	u64 stream_start_position;
+	u32 dai_sample_size;
+	u64 llp1, llp2;
+	u64 wclk1, wclk2;
+	u64 wclk, ratio;
+	u32 node_id;
+	int i;
+
+	spcm = container_of(work, struct snd_sof_pcm_stream, wclk_calibration.work);
+	/* find host & dai to locate info in memory window */
+	for_each_dapm_widgets(spcm->list, i, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+
+		if (!swidget)
+			continue;
+
+		if (WIDGET_IS_AIF(swidget->widget->id)) {
+			host_copier = swidget->private;
+		} else if (WIDGET_IS_DAI(swidget->widget->id)) {
+			dai = swidget->private;
+			dai_copier = dai->private;
+		}
+	}
+
+	sdev = spcm->sdev;
+
+	if (!host_copier || !dai_copier) {
+		dev_err(sdev->dev, "failed to find host copier %p or dai copier%p\n",
+			host_copier, dai_copier);
+		return;
+	}
+
+	fw_reg = spcm->fw_reg;
+	sof_mailbox_read(sdev, sdev->fw_info_box.offset, fw_reg, sizeof(*fw_reg));
```

Comment:
Do you really need to whole fw_regs?

Reply:
sure. I need to check abi and search pipeline reg, llp slots. small read doesn't save time